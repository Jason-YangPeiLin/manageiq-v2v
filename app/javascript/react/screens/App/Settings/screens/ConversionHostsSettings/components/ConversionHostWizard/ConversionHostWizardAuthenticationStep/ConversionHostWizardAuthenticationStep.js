import React from 'react';
import { Form } from 'patternfly-react';

const ConversionHostWizardAuthenticationStep = () => (
  <Form className="form-horizontal">
    <h2>TODO: Authentication Step Contents</h2>
  </Form>
);

export default ConversionHostWizardAuthenticationStep;
