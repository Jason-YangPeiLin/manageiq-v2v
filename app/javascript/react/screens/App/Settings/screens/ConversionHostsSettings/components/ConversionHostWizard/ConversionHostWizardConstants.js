export const stepIDs = {
  locationStep: 'conversionHostWizardLocationStep',
  hostsStep: 'conversionHostWizardHostsStep',
  authenticationStep: 'conversionHostWizardAuthenticationStep',
  resultsStep: 'conversionHostWizardResultsStep'
};
