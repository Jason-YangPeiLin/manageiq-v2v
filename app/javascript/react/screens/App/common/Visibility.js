import React from 'react';
import PropTypes from 'prop-types';
import { noop } from 'patternfly-react';

// This component has no role on the page, and will be hidden from screen readers (role="presentation").
// Renders the children wrapped in visibility: hidden if the condition is met.
// Either way, the children take up space, visibility will not change the flow of the layout.
// The onKeyUp handler and role are used to satisfy eslint-plugin-jsx-a11y.
const Visibility = ({ children, hidden }) => (
  <span style={{ visibility: hidden ? 'hidden' : '' }} onKeyUp={noop} role="presentation">
    {children}
  </span>
);

Visibility.propTypes = {
  children: PropTypes.node,
  hidden: PropTypes.bool
};

export default Visibility;
