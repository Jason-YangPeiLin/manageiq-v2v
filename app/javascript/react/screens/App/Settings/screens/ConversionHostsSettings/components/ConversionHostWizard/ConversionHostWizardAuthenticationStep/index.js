import ConversionHostWizardAuthenticationStep from './ConversionHostWizardAuthenticationStep';

export default ConversionHostWizardAuthenticationStep;
