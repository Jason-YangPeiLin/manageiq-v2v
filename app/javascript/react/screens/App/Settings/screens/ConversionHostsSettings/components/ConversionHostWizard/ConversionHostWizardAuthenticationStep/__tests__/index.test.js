import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { reducer as formReducer, change } from 'redux-form';

import { generateStore } from '../../../../../../../common/testReduxHelpers';
import ConversionHostWizardAuthenticationStep from '../ConversionHostWizardAuthenticationStep';
import ConversionHostWizardAuthenticationStepContainer from '../index';
import { OPENSTACK } from '../../../../../../../../../../common/constants';
import { stepIDs } from '../../ConversionHostWizardConstants';

describe('ConversionHostWizardAuthenticationStep integration test', () => {
  const mountComponentWithNewStore = () => {
    const store = generateStore(
      { form: formReducer },
      { form: { conversionHostWizardLocationStep: { values: { providerType: OPENSTACK } } } }
    );
    const wrapper = mount(
      <Provider store={store}>
        <ConversionHostWizardAuthenticationStepContainer />
      </Provider>
    );
    return { wrapper, store };
  };

  it('should mount ConversionHostWizardAuthenticationStep with mapStateToProps reduced', () => {
    const { wrapper } = mountComponentWithNewStore();

    // query the unconnected component to assert reduced props
    const component = wrapper.find(ConversionHostWizardAuthenticationStep);
    expect(component.props()).toMatchSnapshot();
  });

  it('fails validation when there are empty required fields', () => {
    const { store } = mountComponentWithNewStore();
    expect(store.getState().form[stepIDs.authenticationStep].syncErrors).toBeTruthy();
  });

  // This test case is related to the bug fixed in https://github.com/ManageIQ/manageiq-v2v/pull/961
  it('passes validation when required fields are unmounted while empty', () => {
    const { store } = mountComponentWithNewStore();
    const form = () => store.getState().form[stepIDs.authenticationStep];
    const changeFormValue = (key, val) => store.dispatch(change(stepIDs.authenticationStep, key, val));

    // Form starts with two empty required fields mounted. (with VDDK default selected)
    expect(Object.keys(form().syncErrors)).toEqual(['conversionHostSshKey', 'vddkLibraryPath']);

    // Filling in conversionHostSshKey and switching to SSH transformation mounts another empty required field (vmwareSshKey).
    changeFormValue('conversionHostSshKey', { filename: '', body: 'foo' });
    changeFormValue('transformationMethod', 'SSH');
    expect(Object.keys(form().syncErrors)).toEqual(['vmwareSshKey']); // The vddkLibraryPath that was unmounted is still empty+required, but has no error.

    // Filling in vmwareSshKey makes the form pass validation.
    changeFormValue('vmwareSshKey', { filename: '', body: 'foo' });
    expect(form().syncErrors).toBeFalsy();

    // Changing the transformation method mounts an empty required field (vddkLibraryPath).
    changeFormValue('transformationMethod', 'VDDK');
    expect(Object.keys(form().syncErrors)).toEqual(['vddkLibraryPath']);

    // Changing the transformation method back unmounts vddkLibraryPath while it's empty
    // and remounts vmwareSshKey with its value from earlier, making the form pass validation.
    changeFormValue('transformationMethod', 'SSH');
    expect(form().syncErrors).toBeFalsy(); // <-------

    // Turning on OpenStack TLS Certificates mounts an empty required field (openstackCaCerts)
    changeFormValue('verifyOpenstackCerts', true);
    expect(Object.keys(form().syncErrors)).toEqual(['openstackCaCerts']);

    // Turning it off unmounts openstackCaCerts while it's empty.
    changeFormValue('verifyOpenstackCerts', false);
    expect(form().syncErrors).toBeFalsy(); // <-------
  });
});
