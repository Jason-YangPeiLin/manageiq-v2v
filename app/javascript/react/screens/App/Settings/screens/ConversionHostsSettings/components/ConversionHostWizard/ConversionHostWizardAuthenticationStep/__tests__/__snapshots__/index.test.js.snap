// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversionHostWizardAuthenticationStep integration test should mount ConversionHostWizardAuthenticationStep with mapStateToProps reduced 1`] = `
Object {
  "selectedProviderType": "openstack",
  "selectedTransformationMethod": "VDDK",
  "verifyOpenstackCerts": false,
}
`;
